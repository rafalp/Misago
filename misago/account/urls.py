from django.urls import path

from .views import settings, validate


urlpatterns = [
    path("", settings.index, name="account-settings"),
    path(
        "preferences/",
        settings.AccountPreferencesView.as_view(),
        name="account-preferences",
    ),
    path(
        "details/",
        settings.AccountDetailsView.as_view(),
        name="account-details",
    ),
    path(
        "username/",
        settings.AccountUsernameView.as_view(),
        name="account-username",
    ),
    path(
        "password/",
        settings.AccountPasswordView.as_view(),
        name="account-password",
    ),
    path(
        "email/",
        settings.AccountEmailView.as_view(),
        name="account-email",
    ),
    path(
        "email/confirm/",
        settings.AccountEmailConfirm.as_view(),
        name="account-email-confirm-sent",
    ),
    path(
        "email/confirm/<int:user_id>/<token>/",
        settings.account_email_confirm_change,
        name="account-email-confirm-change",
    ),
    path(
        "download-data/",
        settings.AccountDownloadDataView.as_view(),
        name="account-download-data",
    ),
    path(
        "delete/",
        settings.AccountDeleteView.as_view(),
        name="account-delete",
    ),
    path(
        "delete/completed/",
        settings.account_delete_completed,
        name="account-delete-completed",
    ),
    path(
        "validate/email/",
        validate.email,
        name="account-validate-email",
    ),
    path(
        "validate/username/",
        validate.username,
        name="account-validate-username",
    ),
    path(
        "validate/password/",
        validate.password,
        name="account-validate-password",
    ),
]
