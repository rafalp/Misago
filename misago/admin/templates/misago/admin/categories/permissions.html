{% extends "misago/admin/generic/permissions_form.html" %}
{% load i18n %}


{% block title %}
{% trans "Category permissions" context "admin category form" %} | {{ target }} | {{ active_link.name }} | {{ block.super }}
{% endblock title %}


{% block page-header %}
{{ block.super }}
<small>
  {{ target }}
</small>
{% endblock page-header %}


{% block page-actions %}
<div class="col-auto page-action">
  <a href="{% url 'misago:admin:categories:edit' pk=target.pk %}" class="btn btn-primary btn-sm">
    <span class="fa fa-edit"></span>
    {% trans "Edit category" context "admin categories" %}
  </a>
</div>
<div class="col-auto page-action">
  <a href="{{ target.get_absolute_url }}" class="btn btn-primary btn-sm">
    <span class="fas fa-external-link-alt"></span>
    {% trans "View on site" context "admin categories" %}
  </a>
</div>
{% endblock %}


{% block form-header %}
{% trans "Category permissions" context "admin category form" %}
{% endblock %}


{% block item-header %}
{% trans "Category" context "admin category form" %}
{% endblock item-header %}