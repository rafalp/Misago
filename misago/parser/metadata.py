from typing import Iterable

from django.conf import settings
from django.contrib.auth import get_user_model

from ..core.utils import slugify
from .context import ParserContext
from .hooks import (
    get_ast_metadata_users_queryset_hook,
    update_ast_metadata_hook,
    update_ast_metadata_from_node_hook,
    update_ast_metadata_users_hook,
)

User = get_user_model()


def create_ast_metadata(
    context: ParserContext,
    ast: list[dict],
) -> dict:
    metadata = {
        "outbound-links": set(),
        "usernames": set(),
        "users": {},
        "posts": {
            "ids": set(),
            "objs": {},
        },
    }

    return update_ast_metadata_hook(_update_ast_metadata_action, context, ast, metadata)


def _update_ast_metadata_action(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
) -> dict:
    for ast_node in ast:
        update_ast_metadata_from_node(context, ast_node, metadata)

    update_ast_metadata_users(context, metadata)

    return metadata


def update_ast_metadata_from_node(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    update_ast_metadata_from_node_hook(
        _update_ast_metadata_from_node_action, context, ast_node, metadata
    )


def _update_ast_metadata_from_node_action(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    if ast_node["type"] == "mention":
        metadata["usernames"].add(slugify(ast_node["username"]))

    elif ast_node["type"] == "quote-bbcode":
        if ast_node["author"]:
            metadata["usernames"].add(slugify(ast_node["author"]))
        if ast_node["post"]:
            metadata["posts"]["ids"].add(ast_node["post"])

    elif ast_node["type"] in ("auto-link", "auto-url", "url", "url-bbcode"):
        if not context.forum_address.is_inbound_link(ast_node["href"]):
            metadata["outbound-links"].add(ast_node["href"])

    if ast_node.get("children"):
        for child_node in ast_node["children"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("items"):
        for child_node in ast_node["items"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("lists"):
        for child_node in ast_node["lists"]:
            update_ast_metadata_from_node(context, child_node, metadata)


def update_ast_metadata_users(context: ParserContext, metadata: dict) -> None:
    update_ast_metadata_users_hook(_update_ast_metadata_users_action, context, metadata)


def _update_ast_metadata_users_action(context: ParserContext, metadata: dict) -> None:
    if not metadata["usernames"]:
        return

    usernames = sorted(metadata["usernames"])
    if len(usernames) > settings.MISAGO_PARSER_MAX_USERS:
        return

    if usernames:
        queryset = get_ast_metadata_users_queryset(context, usernames)
        for user in queryset:
            metadata["users"][user.slug] = user


def get_ast_metadata_users_queryset(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return get_ast_metadata_users_queryset_hook(
        _get_ast_metadata_users_queryset_action, context, usernames
    )


def _get_ast_metadata_users_queryset_action(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return User.objects.filter(slug__in=usernames)
