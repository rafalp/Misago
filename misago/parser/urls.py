def clean_href(href: str) -> str:
    if href.startswith("://"):
        return "http" + href
    if "://" not in href:
        return "http://" + href
    return href
