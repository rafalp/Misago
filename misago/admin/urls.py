from django.urls import path

from .. import admin
from .views import auth, index, select

urlpatterns = [
    # "misago:admin:index" link symbolises "root" of Misago admin links space
    # any request with path that falls below this one is assumed to be directed
    # at Misago Admin and will be checked by Misago Admin Middleware
    path("", index.admin_index, name="index"),
    path("select/user/", select.select_user, name="select-user"),
    path("logout/", auth.logout, name="logout"),
]

# Discover admin and register patterns
admin.discover_misago_admin()
urlpatterns += admin.urlpatterns()
