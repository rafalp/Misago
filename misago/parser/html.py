from html import escape

from django.utils.translation import pgettext

from ..core.utils import slugify
from .context import ParserContext
from .exceptions import AstError
from .hooks import complete_markup_html_hook, render_ast_node_to_html_hook
from .urls import clean_href

SPOILER_SUMMARY = "<spoiler-summary-message>"


def render_ast_to_html(context: ParserContext, ast: list[dict], metadata: dict) -> str:
    html = ""
    for ast_node in ast:
        node_html = render_ast_node_to_html(context, ast_node, metadata)
        if node_html:
            html += node_html
    return html


def render_ast_node_to_html(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    return render_ast_node_to_html_hook(
        _render_ast_node_to_html_action, context, ast_node, metadata
    )


def _render_ast_node_to_html_action(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    ast_type = ast_node["type"]

    if ast_type in ("heading", "heading-setex"):
        html_tag = f"h{ast_node['level']}"
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list":
        html_tag = "ol" if ast_node["ordered"] else "ul"
        children = render_ast_to_html(context, ast_node["items"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list-item":
        text = render_ast_to_html(context, ast_node["children"], metadata)
        children = render_ast_to_html(context, ast_node["lists"], metadata)
        return f"<li>{text}{children}</li>"

    if ast_type in ("code", "code-bbcode"):
        if ast_node["syntax"]:
            html_class = f" class=\"language-{ast_node['syntax']}\""
        else:
            html_class = ""
        return f"<pre{html_class}><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-indented":
        return f"<pre><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-inline":
        return f"<code>{escape(ast_node['code'])}</code>"

    if ast_type == "quote":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<blockquote>{children}</blockquote>"

    if ast_type == "quote-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)

        if not ast_node["author"]:
            return f"<blockquote>{children}</blockquote>"

        heading = escape(ast_node["author"])

        return (
            '<aside class="quote-block">'
            f'<div class="quote-heading" data-noquote="1">{heading}</div>'
            f'<blockquote class="quote-body">{children}</blockquote>'
            "</aside>"
        )

    if ast_type == "spoiler-bbcode":
        if ast_node["summary"]:
            summary = escape(ast_node["summary"])
        else:
            summary = SPOILER_SUMMARY
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<details><summary>{summary}</summary>{children}</details>"

    if ast_type == "paragraph":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<p>{children}</p>"

    if ast_type in ("emphasis", "emphasis-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<em>{children}</em>"

    if ast_type in ("strong", "strong-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<strong>{children}</strong>"

    if ast_type in ("strikethrough", "strikethrough-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<del>{children}</del>"

    if ast_type == "bold-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<b>{children}</b>"

    if ast_type == "italics-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<i>{children}</i>"

    if ast_type == "underline-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<u>{children}</u>"

    if ast_type == "superscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sup>{children}</sup>"

    if ast_type == "subscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sub>{children}</sub>"

    if ast_type in ("thematic-break", "thematic-break-bbcode"):
        return "<hr />"

    if ast_type in ("image", "image-bbcode"):
        src = escape(clean_href(ast_node["src"]))
        alt = escape(ast_node["alt"]) if ast_node["alt"] else ""
        return f'<img src="{src}" alt="{alt}" />'

    if ast_type in ("url", "url-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        return f'<a href="{href}" rel="{rel}" target="_blank">{children or href}</a>'

    if ast_type in ("auto-link", "auto-url"):
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        if ast_node.get("image"):
            return f'<img src="{href}" alt="" />'
        else:
            return f'<a href="{href}" rel="{rel}" target="_blank">{href}</a>'

    if ast_type == "mention":
        username = slugify(ast_node["username"])
        if username not in metadata["users"]:
            return escape("@" + ast_node["username"])

        user = metadata["users"][username]
        return f'<a href="{user.get_absolute_url()}">@{escape(user.username)}</a>'

    if ast_type == "escape":
        return escape(ast_node["character"])

    if ast_type == "line-break":
        return "<br />"

    if ast_type == "text":
        return escape(ast_node["text"])

    raise AstError(f"Unknown AST node type: {ast_type}")


def complete_markup_html(html: str, **kwargs) -> str:
    return complete_markup_html_hook(_complete_markup_html_action, html, **kwargs)


def _complete_markup_html_action(html: str, **kwargs) -> str:
    html = complete_markup_html_spoiler_summary(html)
    return html


def complete_markup_html_spoiler_summary(html: str) -> str:
    if SPOILER_SUMMARY in html:
        html = html.replace(
            SPOILER_SUMMARY, pgettext("spoiler summary", "Reveal spoiler")
        )

    return html
