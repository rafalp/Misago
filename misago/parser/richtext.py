import re
from functools import partial

from django.template.loader import render_to_string
from django.utils.translation import pgettext

from ..attachments.models import Attachment
from ..permissions.checkutils import PermissionCheckResult
from .hooks import replace_rich_text_tokens_hook


def replace_rich_text_tokens(html: str, data: dict | None = None) -> str:
    if data is None:
        data = {}

    return replace_rich_text_tokens_hook(_replace_rich_text_tokens_action, html, data)


def _replace_rich_text_tokens_action(html: str, data) -> str:
    html = replace_rich_text_tokens_attachments(
        html, data.get("attachments"), data.get("attachment_errors")
    )

    return html


ATTACHMENT_TOKEN = re.compile(r"\<attachment=(.+?)\>")


def replace_rich_text_tokens_attachments(
    html: str,
    attachments: dict[int, Attachment] | None,
    attachment_errors: dict[int, PermissionCheckResult],
) -> str:
    attachments = attachments or {}
    attachment_errors = attachment_errors or {}

    replace = partial(
        replace_rich_text_attachment_token, attachments, attachment_errors
    )

    return ATTACHMENT_TOKEN.sub(replace, html)


def replace_rich_text_attachment_token(
    attachments: dict[int, Attachment],
    attachment_errors: dict[int, PermissionCheckResult],
    matchobj: re.Match,
) -> str:
    name, slug, id = matchobj.group(1).split(":")

    id = int(id)
    name = name.strip()
    slug = slug.strip()

    error = attachment_errors.get(id)
    if error is not None and error.permission_denied:
        return render_to_string(
            "misago/rich_text/attachment_permission_denied.html",
            {
                "error": error.error,
                "name": name,
            },
        )

    if attachment := attachments.get(id):
        if not attachment.upload:
            template_name = "misago/rich_text/attachment_broken.html"
        elif attachment.filetype.is_image:
            template_name = "misago/rich_text/attachment_image.html"
        elif attachment.filetype.is_video:
            template_name = "misago/rich_text/attachment_video.html"
        else:
            template_name = "misago/rich_text/attachment_file.html"

        return render_to_string(template_name, {"attachment": attachment})

    return render_to_string(
        "misago/rich_text/attachment_link.html",
        {
            "name": name,
            "slug": slug,
            "id": id,
        },
    )
