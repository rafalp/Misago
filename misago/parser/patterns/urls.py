import re
from functools import cached_property

from django.utils.crypto import get_random_string

from ..parents import has_invalid_parent
from ..parser import Parser, Pattern

MAIL_RE = re.compile(r"^\w+[.-_+\w]*@[-\w]+(.\w+)+$")
URL_RE = re.compile(
    r"^(((https?)|(ftps?))://)?\w+(([-_\w])?\w+)*(\.\w+(([-_\w])?\w+))+[^\s]*$"
)


def clean_url(value: str) -> str | None:
    value = value.strip()
    if not value:
        return None

    if MAIL_RE.match(value):
        return f"mailto:{value}"

    if URL_RE.match(value):
        return value

    return None


BBCODE_CONTENTS = re.compile(
    r"\[[a-z]+(=(?P<arg>(.+)))?\](?P<content>((.|\n)*?))\[\/[a-z]+\]", re.IGNORECASE
)


class UrlBBCode(Pattern):
    pattern_type: str = "url-bbcode"
    pattern: str = r"((\[url=.+?\](.|\n)*?)|(\[url\].*?))\[\/url\]"
    invalid_parents: set[str] = {pattern_type, "url"}

    def parse(self, parser: Parser, match: str, parents: list[dict]) -> dict:
        contents = BBCODE_CONTENTS.match(match).groupdict()

        arg = contents["arg"].strip(" \"'") if contents["arg"] else None
        content = contents["content"]

        if arg:
            url = clean_url(arg)
        elif content:
            url = clean_url(content.strip())
            content = None
        else:
            url = None

        if (
            not url
            or not contents["content"].strip()
            or has_invalid_parent(self.invalid_parents, parents)
        ):
            return {"type": "text", "text": match}

        return self.make_ast(parser, match, url, content, parents)

    def make_ast(
        self,
        parser: Parser,
        match: str,
        url: str | None,
        content: str,
        parents: list[dict],
    ) -> dict:
        if content:
            children = parser.parse_inline(content, parents + [self.pattern_type])
        else:
            children = []

        return {
            "type": self.pattern_type,
            "href": url,
            "children": children,
        }


class ImgBBCode(UrlBBCode):
    pattern_type: str = "image-bbcode"
    pattern: str = r"((\[img=.+?\](.|\n)*?)|(\[img\].*?))\[\/img\]"
    invalid_parents: set[str] = {pattern_type, "image"}

    def make_ast(
        self,
        parser: Parser,
        match: str,
        url: str | None,
        alt_text: str | None,
        parents: list[dict],
    ) -> dict:
        if url.startswith("mailto:"):
            return {"type": "text", "text": match}

        if alt_text:
            alt_text = alt_text.strip()

        return {
            "type": self.pattern_type,
            "alt": alt_text or None,
            "src": url,
        }


INLINE_CODE_PATTERN = r"(?<!\\)`(.|\n)*?(?<!\\)`"

IMAGE_PATTERN = r"!(\[(.|\n)*?\])?\(.*?\)"
IMAGE_RE = re.compile(IMAGE_PATTERN)
URL_START_RE = re.compile(r"\[(.|\n)*?\]\(")


class UrlMarkdown(Pattern):
    pattern_type: str = "url"
    pattern: str = r"(?<!\!)\[.*\]\(.*\)"
    exclude_patterns: list[str] = [INLINE_CODE_PATTERN, IMAGE_PATTERN]

    def parse(self, parser: Parser, match: str, parents: list[dict]) -> list[dict]:
        clean_match, reserved_patterns = self.prepare_match_str(match)
        return self.parse_clean_match(parser, clean_match, reserved_patterns, parents)

    def parse_clean_match(
        self,
        parser: Parser,
        clean_match: str,
        reserved_patterns: dict[str, str],
        parents: list[dict],
    ) -> list[dict]:
        ast: list[dict] = []
        cursor = 0

        urls_starts = list(URL_START_RE.finditer(clean_match))
        urls_starts_last = len(urls_starts) - 1
        for i, m in enumerate(urls_starts):
            if m.start() > cursor:
                markup = clean_match[cursor : m.start()]
                if reserved_patterns:
                    markup = self.reverse_text_patterns(markup, reserved_patterns)
                ast += parser.parse_inline(markup, parents)

            text = m.group(0)[1:-2]
            if reserved_patterns:
                text = self.reverse_text_patterns(text, reserved_patterns)

            url_source_start = m.end() - 1
            if i < urls_starts_last:
                next_url_start = urls_starts[i + 1].start()
                url_source = clean_match[url_source_start:next_url_start]
            else:
                url_source = clean_match[url_source_start:]

            url, end = self.extract_url_from_source(url_source)
            if text and url:
                ast.append(self.make_link_ast(parser, text, url, parents))
                cursor = m.end() - 1 + end
            else:
                markup = self.reverse_text_patterns(m.group(0), reserved_patterns)
                for ast_node in parser.parse_inline(markup, parents):
                    if ast_node["type"] == "text" and ast and ast[-1]["type"] == "text":
                        ast[-1]["text"] += ast_node["text"]
                    else:
                        ast.append(ast_node)

                cursor = m.end()

        if cursor < len(clean_match):
            markup = clean_match[cursor:]
            if reserved_patterns:
                markup = self.reverse_text_patterns(markup, reserved_patterns)
            ast += parser.parse_inline(markup, parents)

        return ast

    def prepare_match_str(self, match: str) -> tuple[str, dict[str, str]]:
        clean_match: str = ""
        reserved_patterns = {}

        cursor = 0
        for m in self._exclude_patterns_re.finditer(match):
            if m.start() > cursor:
                clean_match += match[cursor : m.start()]

            pattern_id = f"%%{get_random_string(8)}%%"
            clean_match += pattern_id
            reserved_patterns[pattern_id] = m.group(0)
            cursor = m.end()

        if cursor < len(match):
            clean_match += match[cursor:]

        return clean_match, reserved_patterns

    def reverse_text_patterns(self, text: str, reserved_patterns: dict) -> str:
        for token, value in reserved_patterns.items():
            text = text.replace(token, value)
        return text

    def extract_url_from_source(self, source: str) -> tuple[str, int]:
        opening = source.count("(")
        closing = source.count(")")

        # Simplest case, one opening parenthesis matched by one closing
        if opening == closing and opening == 1:
            url = source[1 : source.find(")")].strip()
            return url, source.find(")") + 1

        if not closing:
            return "", 0

        # Walk the string to find pairs of parentheses
        nesting = 0
        length = 0
        for i, c in enumerate(source):
            if c == "(":
                nesting += 1
            elif c == ")":
                nesting -= 1
                if nesting >= 0:
                    length = i

        url = source[1:length].strip()
        return url, length + 1

    def make_link_ast(
        self, parser: Parser, text: str, url: str, parents: list[dict]
    ) -> dict:
        return {
            "type": self.pattern_type,
            "href": url,
            "children": parser.parse_inline(text, parents + [self.pattern_type]),
        }

    @cached_property
    def _exclude_patterns_re(self) -> re.Pattern:
        return re.compile("|".join(f"({p})" for p in self.exclude_patterns))


IMAGE_CONTENTS = re.compile(r"!(\[(?P<alt>(.|\n)*?)\])?\((?P<src>.*?)\)")


class ImgMarkdown(Pattern):
    pattern_type: str = "image"
    pattern: str = IMAGE_PATTERN

    def parse(self, parser: Parser, match: str, parents: list[dict]) -> list[dict]:
        contents = IMAGE_CONTENTS.match(match).groupdict()

        if contents["alt"]:
            alt = contents["alt"].strip() or None
        else:
            alt = None

        src = contents["src"].strip()
        if URL_RE.match(src):
            return {
                "type": "image",
                "alt": alt,
                "src": src,
            }

        return {"type": "text", "text": match}


class AutolinkMarkdown(Pattern):
    pattern_type: str = "auto-link"
    pattern: str = r"\<!?[^\s]*?\>"
    invalid_parents: set[str] = {UrlMarkdown.pattern_type, UrlBBCode.pattern_type}

    def parse(self, parser: Parser, match: str, parents: list[dict]) -> dict:
        url = match[1:-1]
        if url and url[0] == "!":
            url = url[1:]
            is_image = True
        else:
            is_image = False

        url = clean_url(url)

        if (
            not url
            or (is_image and url.startswith("mailto:"))
            or has_invalid_parent(self.invalid_parents, parents)
        ):
            return {"type": "text", "text": match}

        return {"type": self.pattern_type, "image": is_image, "href": url}


class AutoUrl(AutolinkMarkdown):
    pattern_type: str = "auto-url"
    pattern: str = (
        r"(?<!\w)((https?://(www\.)?)|(www\.))\w+(-|\w)*(\.\w+(-|\w)*)+[^\s)]*"
    )

    def parse(self, parser: Parser, match: str, parents: list[dict]) -> dict:
        url = clean_url(match)
        if not url or has_invalid_parent(self.invalid_parents, parents):
            return {"type": "text", "text": match}

        return {"type": self.pattern_type, "href": url}
