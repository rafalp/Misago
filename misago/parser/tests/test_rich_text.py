from django.core.exceptions import PermissionDenied
from django.http import Http404
from django.urls import reverse

from ...permissions.checkutils import PermissionCheckResult
from ..html import render_ast_to_html
from ..metadata import create_ast_metadata
from ..richtext import replace_rich_text_tokens


def test_replace_rich_text_tokens_replaces_default_spoiler_summary(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("[spoiler]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    html = render_ast_to_html(parser_context, ast, metadata)
    assert snapshot == replace_rich_text_tokens(html)


def test_replace_rich_text_tokens_replaces_attachment_with_image(image_attachment):

    html = f"<attachment={image_attachment.name}:{image_attachment.slug}:{image_attachment.id}>"
    data = {
        "attachment_errors": {},
        "attachments": {image_attachment.id: image_attachment},
    }

    result = replace_rich_text_tokens(html, data)
    assert f'href="{image_attachment.get_details_url()}"' in result
    assert f'href="{image_attachment.get_absolute_url()}"' in result
    assert f'src="{image_attachment.get_absolute_url()}"' in result


def test_replace_rich_text_tokens_replaces_attachment_with_thumbnail(
    image_thumbnail_attachment,
):
    html = f"<attachment={image_thumbnail_attachment.name}:{image_thumbnail_attachment.slug}:{image_thumbnail_attachment.id}>"
    data = {
        "attachment_errors": {},
        "attachments": {image_thumbnail_attachment.id: image_thumbnail_attachment},
    }

    result = replace_rich_text_tokens(html, data)
    assert f'href="{image_thumbnail_attachment.get_details_url()}"' in result
    assert f'href="{image_thumbnail_attachment.get_absolute_url()}"' in result
    assert f'src="{image_thumbnail_attachment.get_thumbnail_url()}"' in result


def test_replace_rich_text_tokens_replaces_attachment_with_video(video_attachment):
    html = f"<attachment={video_attachment.name}:{video_attachment.slug}:{video_attachment.id}>"
    data = {
        "attachment_errors": {},
        "attachments": {video_attachment.id: video_attachment},
    }

    result = replace_rich_text_tokens(html, data)
    assert f'href="{video_attachment.get_details_url()}"' in result
    assert f'src="{video_attachment.get_absolute_url()}"' in result


def test_replace_rich_text_tokens_replaces_attachment_with_file_link(text_attachment):
    html = f"<attachment={text_attachment.name}:{text_attachment.slug}:{text_attachment.id}>"
    data = {
        "attachment_errors": {},
        "attachments": {text_attachment.id: text_attachment},
    }

    result = replace_rich_text_tokens(html, data)
    assert text_attachment.name in result
    assert f'href="{text_attachment.get_details_url()}"' in result
    assert f'href="{text_attachment.get_absolute_url()}"' in result


def test_replace_rich_text_tokens_replaces_attachment_without_upload_with_broken_link(
    broken_image_attachment,
):
    html = f"<attachment={broken_image_attachment.name}:{broken_image_attachment.slug}:{broken_image_attachment.id}>"
    data = {
        "attachment_errors": {},
        "attachments": {broken_image_attachment.id: broken_image_attachment},
    }

    result = replace_rich_text_tokens(html, data)
    assert broken_image_attachment.name in result
    assert f'href="{broken_image_attachment.get_details_url()}"' not in result
    assert f'href="{broken_image_attachment.get_absolute_url()}"' not in result


def test_replace_rich_text_tokens_replaces_permission_denied_attachment_with_error():
    error = PermissionCheckResult(PermissionDenied("Lorem ipsum dolor met"))
    download_url = reverse(
        "misago:attachment-download", kwargs={"slug": "image-png", "id": 123}
    )
    details_url = reverse(
        "misago:attachment-details", kwargs={"slug": "image-png", "id": 123}
    )

    html = "<attachment=image.png:image-png:123>"
    data = {"attachment_errors": {123: error}, "attachments": {}}

    result = replace_rich_text_tokens(html, data)
    assert "image.png" in result
    assert "Lorem ipsum dolor met" in result
    assert f'href="{download_url}"' not in result
    assert f'href="{details_url}"' not in result


def test_replace_rich_text_tokens_replaces_not_found_attachment_with_link():
    error = PermissionCheckResult(Http404())
    download_url = reverse(
        "misago:attachment-download", kwargs={"slug": "image-png", "id": 123}
    )
    details_url = reverse(
        "misago:attachment-details", kwargs={"slug": "image-png", "id": 123}
    )

    html = "<attachment=image.png:image-png:123>"
    data = {"attachment_errors": {123: error}, "attachments": {}}

    result = replace_rich_text_tokens(html, data)
    assert "image.png" in result
    assert f'href="{download_url}"' in result
    assert f'href="{details_url}"' in result


def test_replace_rich_text_tokens_replaces_not_existing_attachment_with_link():
    download_url = reverse(
        "misago:attachment-download", kwargs={"slug": "image-png", "id": 123}
    )
    details_url = reverse(
        "misago:attachment-details", kwargs={"slug": "image-png", "id": 123}
    )

    html = "<attachment=image.png:image-png:123>"
    data = {"attachment_errors": {}, "attachments": {}}

    result = replace_rich_text_tokens(html, data)
    assert "image.png" in result
    assert f'href="{download_url}"' in result
    assert f'href="{details_url}"' in result
