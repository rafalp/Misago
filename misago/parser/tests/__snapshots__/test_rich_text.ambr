# serializer version: 1
# name: test_replace_rich_text_tokens_replaces_default_spoiler_summary
  '<spoiler><summary></summary><body><p>Hello world!</p></body></spoiler>'
# ---
